[bc-breaking][quant][fx][graphmode] Add backend_config_dict for standalone module

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #70152
* #70287
* #70286
* #70277
* #70151
* __->__ #70150

Summary:
This PR allows user to specify backend_config_dict for standalone modules, both in prepare and convert step
adding this now to allow prototype for some of our customer use cases, test for the codepath will be added in
a separate PR

BC-breaking
This adds a new field to the standalone_module configurations (can be None):

Before:
```
prepare_custom_config_dict = {
    "standalone_module_name": [("standalone", sm_qconfig_dict, sm_prepare_custom_config_dict )]
}
m = prepare_fx(m, qconfig_dict, prepare_custom_config_dict=prepare_custom_config_dict
```
After:
```
prepare_custom_config_dict = {
    "standalone_module_name": [("standalone", sm_qconfig_dict, sm_prepare_custom_config_dict, backend_config_dict )]  # backend_config_dict can be None, in which case we'll fall back to use backend_config_dict of parent
}
m = prepare_fx(m, qconfig_dict, prepare_custom_config_dict=prepare_custom_config_dict
```

Test Plan:
regression tests
```
python test/test_quantization.py TestQuantizeFx
```
test that specifies backend_config for some module will be added in a separate PR for the use case we have in mind
since it requires other features

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33205162](https://our.internmc.facebook.com/intern/diff/D33205162)

jerryzh168 committed:
[quant][fx][graphmode] Add backend_config_dict for standalone module

Summary:
This PR allows user to specify backend_config_dict for standalone modules, both in prepare and convert step
adding this now to allow prototype for some of our customer use cases, test for the codepath will be added in
a separate PR

Test Plan:
regression tests
```
python test/test_quantization.py TestQuantizeFx
```
test that specifies backend_config for some module will be added in a separate PR for the use case we have in mind
since it requires other features

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][graphmode] Add backend_config_dict for standalone module"


Summary:
This PR allows user to specify backend_config_dict for standalone modules, both in prepare and convert step
adding this now to allow prototype for some of our customer use cases, test for the codepath will be added in
a separate PR

Test Plan:
regression tests
```
python test/test_quantization.py TestQuantizeFx
```
test that specifies backend_config for some module will be added in a separate PR for the use case we have in mind
since it requires other features

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33205162).

jerryzh168 committed:
Update on "[quant][fx][graphmode] Add backend_config_dict for standalone module"


Summary:
This PR allows user to specify backend_config_dict for standalone modules, both in prepare and convert step
adding this now to allow prototype for some of our customer use cases, test for the codepath will be added in
a separate PR

Test Plan:
regression tests
```
python test/test_quantization.py TestQuantizeFx
```
test that specifies backend_config for some module will be added in a separate PR for the use case we have in mind
since it requires other features

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33205162](https://our.internmc.facebook.com/intern/diff/D33205162)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33205162).

vkuzo commented on torch/ao/quantization/quantize_fx.py:
```diff
@@ -632,6 +632,22 @@ def convert_fx(
                  }
               },
 
+              # optional: specify the path for standalone modules
```
it's unexpected that both `prepare_fx` and `convert_fx` support different flavors of standalone module, can we converge on just one and only have it in one place?

jerryzh168 replied:
yeah ideally we should do that. currently the module type is not preserved in symbolic tracing, so we need to add that first, I remember we have discussed about this at some point, not sure what is the plan on that, do you remember what is the plan for preserving the types in symbolic tracing?

vkuzo replied:
can the new functionality live in the same places as the old one (either both live in prepare, or both live in convert), and can we add a comment on the restrictions and plans to remove it?

jerryzh168 replied:
I just looked at the code again, and I find that we can remove this since we are supporting the backend_config_dict only in the new convert flow, this config was added initially because I was planning to support this in the current convert function. will remove this part of docs

jerryzh168 committed:
Update on "[quant][fx][graphmode] Add backend_config_dict for standalone module"


Summary:
This PR allows user to specify backend_config_dict for standalone modules, both in prepare and convert step
adding this now to allow prototype for some of our customer use cases, test for the codepath will be added in
a separate PR

Test Plan:
regression tests
```
python test/test_quantization.py TestQuantizeFx
```
test that specifies backend_config for some module will be added in a separate PR for the use case we have in mind
since it requires other features

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33205162](https://our.internmc.facebook.com/intern/diff/D33205162)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][graphmode] Add backend_config_dict for standalone module"


Summary:
This PR allows user to specify backend_config_dict for standalone modules, both in prepare and convert step
adding this now to allow prototype for some of our customer use cases, test for the codepath will be added in
a separate PR

Test Plan:
regression tests
```
python test/test_quantization.py TestQuantizeFx
```
test that specifies backend_config for some module will be added in a separate PR for the use case we have in mind
since it requires other features

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33205162](https://our.internmc.facebook.com/intern/diff/D33205162)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33205162).

jerryzh168 committed:
Update on "[quant][fx][graphmode] Add backend_config_dict for standalone module"


Summary:
This PR allows user to specify backend_config_dict for standalone modules, both in prepare and convert step
adding this now to allow prototype for some of our customer use cases, test for the codepath will be added in
a separate PR

Test Plan:
regression tests
```
python test/test_quantization.py TestQuantizeFx
```
test that specifies backend_config for some module will be added in a separate PR for the use case we have in mind
since it requires other features

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33205162](https://our.internmc.facebook.com/intern/diff/D33205162)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][fx][graphmode] Add backend_config_dict for standalone module"


Summary:
This PR allows user to specify backend_config_dict for standalone modules, both in prepare and convert step
adding this now to allow prototype for some of our customer use cases, test for the codepath will be added in
a separate PR

Test Plan:
regression tests
```
python test/test_quantization.py TestQuantizeFx
```
test that specifies backend_config for some module will be added in a separate PR for the use case we have in mind
since it requires other features

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33205162](https://our.internmc.facebook.com/intern/diff/D33205162)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33205162).

jerryzh168 committed:
Update on "[quant][fx][graphmode] Add backend_config_dict for standalone module"


Summary:
This PR allows user to specify backend_config_dict for standalone modules, both in prepare and convert step
adding this now to allow prototype for some of our customer use cases, test for the codepath will be added in
a separate PR

Test Plan:
regression tests
```
python test/test_quantization.py TestQuantizeFx
```
test that specifies backend_config for some module will be added in a separate PR for the use case we have in mind
since it requires other features

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33205162](https://our.internmc.facebook.com/intern/diff/D33205162)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33205162).

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33205162).

jerryzh168 committed:
Update on "[quant][fx][graphmode] Add backend_config_dict for standalone module"


Summary:
This PR allows user to specify backend_config_dict for standalone modules, both in prepare and convert step
adding this now to allow prototype for some of our customer use cases, test for the codepath will be added in
a separate PR

Test Plan:
regression tests
```
python test/test_quantization.py TestQuantizeFx
```
test that specifies backend_config for some module will be added in a separate PR for the use case we have in mind
since it requires other features

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33205162](https://our.internmc.facebook.com/intern/diff/D33205162)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33205162).

jerryzh168 committed:
Update on "[bc-breaking][quant][fx][graphmode] Add backend_config_dict for standalone module"


Summary:
This PR allows user to specify backend_config_dict for standalone modules, both in prepare and convert step
adding this now to allow prototype for some of our customer use cases, test for the codepath will be added in
a separate PR

BC-breaking
This adds a new field to the standalone_module configurations (can be None):

Before:
```
prepare_custom_config_dict = {
    "standalone_module_name": [("standalone", sm_qconfig_dict, sm_prepare_custom_config_dict )]
}
m = prepare_fx(m, qconfig_dict, prepare_custom_config_dict=prepare_custom_config_dict
```
After:
```
prepare_custom_config_dict = {
    "standalone_module_name": [("standalone", sm_qconfig_dict, sm_prepare_custom_config_dict, backend_config_dict )]  # backend_config_dict can be None, in which case we'll fall back to use backend_config_dict of parent
}
m = prepare_fx(m, qconfig_dict, prepare_custom_config_dict=prepare_custom_config_dict
```

Test Plan:
regression tests
```
python test/test_quantization.py TestQuantizeFx
```
test that specifies backend_config for some module will be added in a separate PR for the use case we have in mind
since it requires other features

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D33205162](https://our.internmc.facebook.com/intern/diff/D33205162)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33205162).
